### PR TITLE
Instead of split, re should parse output from the `ip route` command

### LINKED ISSUE #58377: saltstack network module fails with ecmp routes
### Description of Issue
Salt network module not working when default route has several destinations. 

### Setup
Routing table looks like:
```
default proto bird metric 32 
	nexthop via 1.2.3.4 dev eth1 weight 1 
	nexthop via 2.3.4.5 dev eth2 weight 1 
```
and `salt-call network.routes` fails with:
```
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
ValueError: u'nexthop' does not appear to be an IPv4 or IPv6 network
Traceback (most recent call last):
  File "/usr/bin/salt-call", line 11, in <module>
    salt_call()
  File "/usr/lib/python2.7/site-packages/salt/scripts.py", line 431, in salt_call
    client.run()
  File "/usr/lib/python2.7/site-packages/salt/cli/call.py", line 57, in run
    caller.run()
  File "/usr/lib/python2.7/site-packages/salt/cli/caller.py", line 138, in run
    ret = self.call()
  File "/usr/lib/python2.7/site-packages/salt/cli/caller.py", line 237, in call
    ret['return'] = self.minion.executors[fname](self.opts, data, func, args, kwargs)
  File "/usr/lib/python2.7/site-packages/salt/executors/direct_call.py", line 12, in execute
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/modules/network.py", line 1686, in default_route
    _routes = routes()
  File "/usr/lib/python2.7/site-packages/salt/modules/network.py", line 1643, in routes
    routes_ = _ip_route_linux()
  File "/usr/lib/python2.7/site-packages/salt/modules/network.py", line 514, in _ip_route_linux
    address_mask = convert_cidr(comps[0])
  File "/usr/lib/python2.7/site-packages/salt/modules/network.py", line 1145, in convert_cidr
    cidr = calc_net(cidr)
  File "/usr/lib/python2.7/site-packages/salt/modules/network.py", line 1167, in calc_net
    return salt.utils.network.calc_net(ip_addr, netmask)
  File "/usr/lib/python2.7/site-packages/salt/utils/network.py", line 1087, in calc_net
    return six.text_type(ipaddress.ip_network(ipaddr, strict=False))
  File "/usr/lib/python2.7/site-packages/salt/ext/ipaddress.py", line 150, in ip_network
    address)
ValueError: u'nexthop' does not appear to be an IPv4 or IPv6 network
```
### Versions Report
```
Salt Version:
           Salt: 2019.2.2
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: 1.10.6
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.26.3
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.8.1
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: 3.7.3
         pygit2: 0.26.4
         Python: 2.7.5 (default, Aug  7 2019, 00:51:29)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.7.1908 Core
         locale: UTF-8
        machine: x86_64
        release: 4.17.6-1.el7.elrepo.x86_64
         system: Linux
        version: CentOS Linux 7.7.1908 Core
```

### LINKED ISSUE #60021: [BUG] network.routes produces inconsistent results when `netstat` is unavailable.
**Description**
When netstat is unavailable the `network.routes` module produces incomplete results with incorrect route flags and missing interfaces.

**Setup**
```
yum install net-tools -y
```

**Steps to Reproduce the behavior**
salt-call --local network.routes
yum remove net-tools -y
salt-call --local network.routes

**Expected behavior**
The network.routes output should be (relatively) consistent

**Screenshots**
```
[root@ussz1jump5001 modules]# which netstat
/bin/netstat
[root@ussz1jump5001 modules]# salt-call --local network.routes family=inet
local:
    |_
      ----------
      addr_family:
          inet
      destination:
          0.0.0.0
      flags:
          UG
      gateway:
          100.64.24.145
      interface:
          eth0
      netmask:
          0.0.0.0
    |_
      ----------
      addr_family:
          inet
      destination:
          100.64.24.144
      flags:
          U
      gateway:
          0.0.0.0
      interface:
          eth0
      netmask:
          255.255.255.240
    |_
      ----------
      addr_family:
          inet
      destination:
          169.254.0.0
      flags:
          U
      gateway:
          0.0.0.0
      interface:
          eth0
      netmask:
          255.255.0.0
    |_
      ----------
      addr_family:
          inet
      destination:
          192.168.0.0
      flags:
          UG <-- populated and correct flags
      gateway:
          100.64.24.145<--- populated and correct gateway
      interface:
          eth0<--- populated and correct interface
      netmask:
          255.255.255.0
    |_
      ----------
      addr_family:
          inet
      destination:
          192.168.1.0
      flags:
          UG <-- populated and correct flags
      gateway:
          100.64.24.145 <--- populated and correct gateway
      interface:
          eth0 <--- populated and correct interface
      netmask:
          255.255.255.0
[root@ussz1jump5001 modules]#
```

```
[root@ussz1jump5001 modules]# yum remove -y -q net-tools
[root@ussz1jump5001 modules]# which netstat
/usr/bin/which: no netstat in (/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin)
[root@ussz1jump5001 modules]# salt-call --local network.routes family=inet
local:
    |_
      ----------
      addr_family:
          inet
      destination:
          0.0.0.0
      flags:
          UG
      gateway:
          100.64.24.145
      interface:
          eth0
      netmask:
          0.0.0.0
    |_
      ----------
      addr_family:
          inet
      destination:
          100.64.24.144
      flags:
          U
      gateway:
          0.0.0.0
      interface:
          eth0
      netmask:
          255.255.255.240
    |_
      ----------
      addr_family:
          inet
      destination:
          169.254.0.0
      flags:
          U 
      gateway:
          0.0.0.0
      interface:
          eth0
      netmask:
          255.255.0.0
    |_
      ----------
      addr_family:
          inet
      destination:
          192.168.0.0
      flags:
          U<--- Wrong Flags
      gateway:
          0.0.0.0 <--- Wrong Gateway
      interface:  <--- Missing Interface
      netmask:
          255.255.255.0
    |_
      ----------
      addr_family:
          inet
      destination:
          192.168.1.0
      flags:
          U <--- Wrong Flags
      gateway:
          0.0.0.0 <--- Wrong Gateway
      interface:  <--- Missing Interface
      netmask:
          255.255.255.0

```

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

NOTE: This behavior is consistent up to and including the current `master` branch. 

```
Salt Version:
           Salt: 2019.2.8

Dependency Versions:
           cffi: 1.13.2
       cherrypy: Not Installed
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.31.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Aug 13 2020, 02:51:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: redhat 7.9 Maipo
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-1160.11.1.el7.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 7.9 Maipo

```
</details>

### PR DESCRIPTION
**Pull request by adammike**
### What does this PR do?
This PR changes the _ip_route_linux function to use a regex to parse the route tables rather than using str.split(). 
### What issues does this PR fix or reference?
Fixes: #58377 , #60021

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by adammike** [comment]
This also fixes #60021

**Comment by adammike** [comment]
Here's a link to my regex101 I used to test the regex pattern against a bunch of lines we pulled from boxes in our environment.

https://regex101.com/r/NDBYSr/1

**Comment by garethgreenaway** [comment]
@adammike Thanks for the PR, the change looks good but we'll need a test to get this merged.

**Comment by adammike** [comment]
@garethgreenaway any objection for me adding tests for `_ip_route_linux` and `_netstat_route_linux` to test_network.py? That'd be the only way you can test this specific bug.. Run the same routes through both functions and make sure the output matches.

**Comment by garethgreenaway** [comment]
That sounds like a perfect way to test the changes!

**Comment by adammike** [comment]
@garethgreenaway done

File: salt/modules/network.py [hunk 1]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
+                flags = "UG"
+            else:
+                flags = "U"
 
             ret.append(
                 {
                     "addr_family": "inet",
                     "destination": address_mask["network"],
-                    "gateway": "0.0.0.0",
+                    "gateway": ip_gateway if ip_gateway else "0.0.0.0",
```
Comment (waynew) [inline salt/modules/network.py:565]: **suggestion**: No need to make this a ternary statement, simply use `or`: `ip_gateway or "0.0.0.0"` should work correctly here.

File: salt/modules/network.py [hunk 2]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
             )
         else:
-            address_mask = convert_cidr(comps[0])
-            ip_interface = ""
-            if comps[1] == "dev":
-                ip_interface = comps[2]
+            address_mask = convert_cidr(line.group("network"))
+            ip_interface = line.group("interface")
+            ip_gateway = line.group("gateway")
+            if line.group("gateway"):
```
Comment (waynew) [inline salt/modules/network.py:556]: **suggestion**: this has already been retrieved once.

```suggestion
            if ip_gateway:
```

would be appropriate here.

File: salt/modules/network.py [hunk 3]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
             ret.append(
                 {
                     "addr_family": "inet6",
                     "destination": "::/0",
-                    "gateway": comps[2],
+                    "gateway": ip_gateway,
                     "netmask": "",
                     "flags": "UG",
-                    "interface": ip_interface,
+                    "interface": ip_interface if ip_interface else "",
```
Comment (waynew) [inline salt/modules/network.py:595]: **suggestion**: similar story here: `ip_interface or ""`

File: salt/modules/network.py [hunk 4]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
+            local_address = line.group("gateway") + "/128"
             ret.append(
                 {
                     "addr_family": "inet6",
                     "destination": local_address,
                     "gateway": "::",
                     "netmask": "",
                     "flags": "U",
-                    "interface": ip_interface,
+                    "interface": ip_interface if ip_interface else "",
```
Comment (waynew) [inline salt/modules/network.py:609]: **suggestion**: similar story here: `ip_interface or ""`

File: salt/modules/network.py [hunk 5]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
                 }
             )
         else:
-            address_mask = convert_cidr(comps[0])
-            ip_interface = ""
-            if comps[1] == "dev":
-                ip_interface = comps[2]
+            ip_interface = line.group("interface")
+            ip_gateway = line.group("gateway")
+            if line.group("gateway") is None:
```
Comment (waynew) [inline salt/modules/network.py:615]: **suggestion**: same story here:
```suggestion
            if ip_gateway is None:
```

File: salt/modules/network.py [hunk 6]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
+            else:
+                flags = "UG"
 
             ret.append(
                 {
                     "addr_family": "inet6",
-                    "destination": comps[0],
-                    "gateway": "::",
+                    "destination": line.group("network"),
+                    "gateway": ip_gateway if ip_gateway else "::",
```
Comment (waynew) [inline salt/modules/network.py:624]: **suggestion**: another place where the ternary could just be `ip_gateway or "::"`

File: salt/modules/network.py [hunk 7]
```diff
@@ -522,103 +522,109 @@ def _ip_route_linux():
...
                     "netmask": "",
-                    "flags": "U",
-                    "interface": ip_interface,
+                    "flags": flags,
+                    "interface": ip_interface if ip_interface else "",
```
Comment (waynew) [inline salt/modules/network.py:627]: **suggestion**: another ternary

**Review by waynew — CHANGES_REQUESTED** [review]
Hey, thanks for the PR, looking great! Just a couple of style changes I noticed here.

**Comment by adammike** [comment]
Requested changes have been made

**Review by joechainz — DISMISSED** [review]
Looks great, thanks you for also testing netstat!

**Review by twangboy — DISMISSED** [review]

**Review by Ch3LL — CHANGES_REQUESTED** [review]
This just needs a changelog entry.

**Comment by Ch3LL** [comment]
ping @adammike did you see my comment about adding a changelog for this change. Want to be able to get this merged in for ya :)

**Comment by adammike** [comment]
I did not. I'll take a look at this tonight.

**Comment by Ch3LL** [comment]
bump. Also looks like there is a merge conflict.

**Comment by Ch3LL** [comment]
If you tell me what you want to add to the changelog I can add the commit to the PR.

File: salt/crypt.py [hunk 8]
```diff
@@ -1417,13 +1417,13 @@ def sign_in(self, timeout=60, safe=True, tries=1, channel=None):
...
+                "The Salt Master server's public key did not authenticate! "
                 "The master may need to be updated if it is a version of Salt "
-                "lower than %s, or\n"
+                "lower than %s, or "
                 "If you are confident that you are connecting to a valid Salt "
                 "Master, then remove the master public key and restart the "
-                "Salt Minion.\nThe master public key can be found "
-                "at:\n%s",
+                "Salt Minion. The master public key can be found "
+                "at: %s",
```
Comment (waynew) [inline salt/crypt.py:1426]: **question** is this actually a related change? Seems like this would go better in a different PR.

File: tests/pytests/unit/modules/test_network.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+def test_netstat_route_linux_ipv4(expected_routes):
+    """
+    Tests that _netstat_route_linux can parse ipv4 routes appropriately
+    """
+    netstat_output = [
+        "0.0.0.0         100.64.24.145   0.0.0.0         UG        0 0          0 eth0",
+        "100.64.24.144   0.0.0.0         255.255.255.240 U         0 0          0 eth0",
+        "169.254.0.0     0.0.0.0         255.255.0.0     U         0 0          0 eth0",
+        "192.168.0.0     100.64.24.145   255.255.255.0   UG        0 0          0 eth0",
+        "192.168.1.0     100.64.24.145   255.255.255.0   UG        0 0          0 eth0",
```
Comment (waynew) [inline tests/pytests/unit/modules/test_network.py:99]: **question** Am I misunderstanding something? https://github.com/saltstack/salt/issues/58377 has the following:

```
default proto bird metric 32 
	nexthop via 1.2.3.4 dev eth1 weight 1 
	nexthop via 2.3.4.5 dev eth2 weight 1 
```

Which doesn't look like anything that's being tested for. Is that an incorrect understanding?

**Review by waynew — CHANGES_REQUESTED** [review]

**Comment by waynew** [comment]
@adammike Hey, just a ping on this - we would like to get this merged in, just needs a bit of love to get it across the line!

**Comment by waynew** [comment]
@adammike any plans to get back to this?

**Comment by waynew** [comment]
Hey @adammike is this something you'll have time for any time soon?

**Comment by adammike** [comment]
Sorry, I've recently changed roles and I've been consumed with other things. 

It's my understanding that you need 3 things, is this accurate?

1) Changeling updated
2) salt/crypt.py logging changes removed form this PR (this was actually a mistake)
3) The included tests do not test for a bug outlined in #58377

I'll revert the crypt.py stuff, the tests I included are pretty straight forward. I took ifconfig and netstat output from the same box, and made sure that the output from this module was consistent between either method. There's probably a bunch of different edge cases that aren't specifically included in the tests, but the logic in the code should be resilient enough handle these. The main goal here was to ensure that the output returned after parsing ip route was consistent with the output returned when obtaining the routes via netstat.

File: salt/crypt.py (hunk 8 above)
Comment (adammike) [reply salt/crypt.py]: removed from this PR

File: tests/pytests/unit/modules/test_network.py (hunk 9 above)
Comment (adammike) [reply tests/pytests/unit/modules/test_network.py:99]: It's been awhile since I looked at this, but I believe when I pasted in the output from that code into my issue, it parsed properly without any errors. 

I didn't add a test for that specific issue, but the core of the issue was poor parsing of ip route output. This code should be much more resilient and less prone to misalignment or keyword mismatches.

**Comment by adammike** [comment]
@Ch3LL the changeling should read: 

Resolved an issue in the `network` execution module  that caused routes detected with the `ip route` command to be inconsistent with routes detected with the `netstat` command on Linux operating systems

I was trying to follow the instructions to update the changeling, but it's been so log I wasn't sure what version to use :-)

**Comment by Ch3LL** [comment]
I added the changelog :)
ping @waynew can you re-review

File: tests/pytests/unit/modules/test_network.py (hunk 9 above)
Comment (waynew) [reply tests/pytests/unit/modules/test_network.py:99]: Hrm. Well, I tried adding the output to both of these tests and they both failed. I even tried providing the output from [this comment](https://github.com/saltstack/salt/issues/58377#issuecomment-800466893) in the ip_route and I got the exact same output in the bug report.

```
    ip_route_output = [
        "::1 dev lo proto kernel metric 256 pref medium",
        "fe80::/64 dev xxx proto kernel metric 100 pref medium",
        "local ::1 dev lo table local proto kernel metric 0 pref medium",
        "local fe80::250:xxxx:xxxx:xxxx dev xxx table local proto kernel metric 0 pref medium",
        "multicast ff00::/8 dev xxx table local proto kernel metric 256 pref medium",
    ]
```
that gives me `ValueError: 'local' does not appear to be an IPv4 or IPv6 network`

and 

```
    ip_route_output = [
         "default proto bird metric 32",
         "   nexthop via 1.2.3.4 dev eth1 weight 1",
         "   nexthop via 2.3.4.5 dev eth2 weight 1",
    ]
```

gives me the same error but `nexthop`.

It's entirely possible that I'm doing it wrong, as I didn't spend an awful lot of time to really grok what the tests/changes are supposed to do, but I'd definitely like to see some tests showing that the mentioned issue is actually resolved.

**Review by waynew — CHANGES_REQUESTED** [review]
